ruby/phase-1.md

```diff
@@ -0,0 +1,536 @@
+
+- There's no need to pollute global scope with activity methods (see "Global Scope" section)
+- There's no need to inject global context for communicating with Temporal. It will be available
+  through a private variable on the class instance (`#activity`)
+- It plays nicely with Ruby's OOP model
+- It allows for future extensibility including 3rd party extensions
+- Activity can be easily tested by injecting a test context into the class instead of the real one
+
+By default we'll use the name of the class as activity name (when communicating with the server).
+This is of course overridable in case the SDK users would like to use their own naming schema. And
+we'll use the method `#execute` as the primary entrypoint for executing user code.
+
+```ruby
+class MyActivity < Temporal::Activity
+  ManuallyCancelled = Class.new(Temporal::Error)
+
+  def execute(arg_1, arg_2: nil)
+    loop do
+      sleep 1
+      activity.logger.info('❤️')
+      activity.logger.debug("Attempt #{activity.info.attempts}")
+      activity.heartbeat('details')
+    end
+  rescue Temporal::Error::ActivityCancelled
+    raise ManuallyCancelled, 'activity got cancelled'
+  end
+end
+```
+
+Notes:
+
+- `activity` is an instance of `Temporal::Activity::Context`
+- `activity.logger` will use a logging adapter configure via `Temporal::Configuration` (more on
+  this in the 2nd phase)
+- Common logging context can be configured by using middleware (more on this in the 2nd phase)
+- `activity.heartbeat` is a synchronous call
+- Errors defined within the activity class will be namespaces `MyActivity::ManuallyCancelled`,
+  which is a very useful property
+
+```ruby
+class Temporal::Activity::Context
+  # Access to a pre-configured logger
+  def logger() -> Temporal::Logger
+
+  # Heartbeat will raise if a cancellation was requested
+  def heartbeat(*details) -> void
+
+  # Provides access to extra details from PollActivityTaskQueueResponse
+  def info() -> Temporal::Info::Activity
+end
+```
+
+
+## Payload converters
+
+Ruby SDK will use the model proposed [here](https://docs.temporal.io/docs/concepts/what-is-a-data-converter/)
+to facilitate conversion between SDK inputs and payloads. Basically there will be two extension
+points on the `Temporal::Configuration` object allowing to add either a DataConverter or a Codec.
+
+The main difference is that a DataConverter translates between any Ruby object and a Payload, while
+a Codec is a lower level construct that encodes/decodes from a Payloads protobuf to a Payloads
+protobuf. The former is intended to support specific serialisation formats, while the later is for
+modifying a binary that is transmitted over the wire.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7933')
+
+config.add_data_converter(MyProtobufConverter.new)
+config.add_data_converter(MyThriftConverter.new)
+
+config.add_codec(EncryptionCodec.new(KEY_ID))
+config.add_codec(CompressionCodec.new)
```

Comment:
Not loving these `add_*` methods.
How do I replace or reorder the default converters for example?
Also, making `Configuration` mutable adds confusion IMHO.

Reply:
Ok, let's try to distill this. We can have codes and converters submitted via initialization of the `Configuration`, like this:

```ruby
Temporal::Configuration.new(
  url: '', 
  data_converters: [MyProtobufConverter.new, MyThriftConverter.new], 
  codecs: [EncryptionCodec.new(KEY_ID), CompressionCodec.new]
)
```

It's a bit more clear with Codecs since there won't be any default ones (I'm assuming), but is it's not clear whether data converters are appended or replace the default converters. We can ask users to always pass the full list of data converters (including the built-in ones), but this means that if we add another one to the SDK, the app users will have to append these manually to their configs to start using them. But maybe it is intended… I'm open to switching to this approach if you think it's an acceptable trade off.

Reply:
I'd just call these `payload_converter` and `payload_codecs`, so there's a single payload converter and the user can compose converters using the framework's CompositePayloadConverter.

Reply:
I don't see a point for the `CompositePayloadConverter` if you almost certainly expected to have multiple payload converters. Do you see the case when you wouldn't use `CompositePayloadConverter`?

This also doesn't solve the problem with missing out on newly added converters since you're overriding the top one (instead of amending these).